217-providers-suggest-order

jwasilgeo opened the pull request:
Resolves enhancement issue #217.

Changes include:
- Now provider suggestion results are housed within a new DOM element child container for each provider, in order to maintain the configured order of providers for suggested results.
- Clearing the parent `this._suggestions` is handled a little bit differently.  Instead of its `innerHTML` being set to `''`, each of the providers' child container is cleared out; again, this is to maintain the original configured order.

Some questions which might create more work for another PR(s):
1. I can't tell how the `clearSuggestions` method is being actively used. Is this a holdover from previous functionality? https://github.com/Esri/esri-leaflet-geocoder/blob/87c9b7ed5335cfba88631b757c41a1fbb3519493/src/Controls/Geosearch.js#L157  It is being called from here but doesn't quite make sense to me. https://github.com/Esri/esri-leaflet-geocoder/blob/87c9b7ed5335cfba88631b757c41a1fbb3519493/src/Classes/GeosearchCore.js#L107
2. **Unrelated to this PR**, there are cases when suggestions should be generated but are not.  This is due to `this._lastValue` being perhaps a little too strict and its value could maybe be set in more places. Consider a user experience when (1) getting suggestions then (2) blurring from the input element then (3) returning to the input element and attempting to get fresh suggestions for the exact same search text.   https://github.com/Esri/esri-leaflet-geocoder/blob/87c9b7ed5335cfba88631b757c41a1fbb3519493/src/Controls/Geosearch.js#L361-L362

gavinr commented:
@jwasilgeo what's a good way to see this in action? What code could/should I add to `debug/sample.html`?

jwasilgeo commented:
Good question.  I would alter the contents of the `providers` config array in the sample page. 
 The order you set here should persist top-to-bottom in the suggestions ui panel (instead of the top-to-bottom order being defined by the order of suggest responses).
https://github.com/Esri/esri-leaflet-geocoder/blob/87c9b7ed5335cfba88631b757c41a1fbb3519493/debug/sample.html#L38

(Also tagging @jgravois for some of my history questions above.)

gavinr commented:
Code looks good to me, and it looks like that basic use case is working, when I change the order of `providers` I get this different behavior:

![image](https://user-images.githubusercontent.com/209355/65437104-d065e280-dde8-11e9-9b06-88f75f3feb19.png)

![image](https://user-images.githubusercontent.com/209355/65437227-0dca7000-dde9-11e9-9f1f-6c2f08ad13f6.png)

Looks like we do not have unit tests or e2e testing around how the DOM looks in the suggestion dropdown, so we will not worry about that for now (although it might be nice to add that in the future).

jgravois commented on src/Controls/Geosearch.js:
```diff
@@ -216,8 +219,15 @@ export var Geosearch = Control.extend({
+    // create a child contents container element for each provider inside of this._suggestions
+    // to maintain the configured order of providers for suggested results
+    for (var i = 0; i < this.options.providers.length; i++) {
+      this.options.providers[i]._contentsElement = DomUtil.create('span', null, this._suggestions);
```
this is just a nit, but semantically I'd expect the container for each provider to be a `<div>` too.

I see above that we've been using a `<span>` for individual headers too though, so no need to go rearranging deck chairs, especially if it makes more .css work.

jwasilgeo replied:
I agree and will give `<div>`s the chance they deserve both with these new elements and the existing ones.

jwasilgeo replied:
Actually worked like a charm without any css changes...yet.  The new structure per each provider is:

```html
<div>
  <div class="geocoder-control-header"></div>
  <ul class="geocoder-control-list"></ul>
<div>
```

jgravois replied:
now that's what I call _semantic_.

jwasilgeo commented:
@gavinr @jgravois thank you both for your reviews and comments!  I'm going to keep this open for now and push some more changes.

jwasilgeo commented:
@gavinr @jgravois I just pushed 3 more commits, each of which I hope have self-explanatory messages. If either of you could review again I'd really appreciate it.

jwasilgeo commented:
@jgravois @gavinr I'm going to write a fresh list of changes now that I added more commits after your all's earlier reviews.  Thoughts?

Changes include:

- Now provider suggestion results are housed within a new DOM element child container for each provider, in order to maintain the configured order of providers for suggested results.  (And using `div` instead of `span` when possible.)
- Clearing the parent `this._suggestions` is handled a little bit differently.  Instead of its `innerHTML` being set to `''`, each of the providers' child container is cleared out; again, this is to maintain the original configured order.
- Moved the hiding of the loading indicator to the end of **all** suggest requests.
- Added more error handling UI cleanup when a provider suggest request fails.
- Wrapped up some repeated lines into methods and removed no longer used lines of code.
- Assigned `this._lastValue` when input text length is too short to perform suggestions on, so that the user can search again with the same text when it gets long enough, e.g. typing and expecting suggestions for "sa" (yes), then "s" (no), then "sa" again (yes).

jgravois commented:
given that @gavinr approved and you addressed my earlier feedback, I'd say 'merge at will'

jwasilgeo commented:
thank you both
